Media: Create media-specific route for stylesheet bundle

jeremeylduvall opened the pull request:
This is a follow-up to #18642. In that PR, we temporarily re-added the media modal style back into the `_components` stylesheet to fix some display issues with the media library in the post editor. This follow-up creates a separate bundle instead for the `/media` and `post-editor` routes specifically. This shouldn't create any visual changes. We're just moving stylesheets around.

I left some inline comments as well for comments.

### To test
1. Run `git checkout update/create-media-css-bundle` and start your local server.
2. Open http://calypso.localhost:3000/media/. Select a site if necessary.
3. Resize the media library, edit an image, upload an image/document/audio file. You shouldn't notice any issues.
4. Open the post editor at http://calypso.localhost:3000/post/. Add a media item to your post. Try to edit that media item by pressing Edit at the top of the media library after selecting an image. You shouldn't notice any regressions.

Fixes: #18641

jeremeylduvall commented on assets/stylesheets/_components.scss:
```diff
@@ -331,7 +331,6 @@
 @import 'my-sites/invites/invite-accept-logged-in/style';
 @import 'my-sites/invites/invite-accept/style';
 @import 'my-sites/media-library/style';
```
If we move out this style for `media-library/style` then clear your cache and load any page in Calypso prior to a section that includes the media routes, you get this "Choose File" input near the Add button:

<img width="606" alt="screen shot 2017-10-17 at 7 50 30 am" src="https://user-images.githubusercontent.com/7240478/31668632-f6d2a31e-b30f-11e7-9a95-18c9e7987a90.png">

As a result, I left the style here in `_components`.

stephanethomas replied:
Good catch. The problem lies with the `MediaLibraryUploadButton` component, which is embedded [in the sidebar](https://github.com/Automattic/wp-calypso/blob/57619e165a42087d6eb111a8c90cd380f4cbf5d5/client/my-sites/sidebar/manage-menu.jsx#L201-L207). We could import `my-sites/media-library/style` from the new `media` CSS bundle, and just import the [stylesheet](https://github.com/Automattic/wp-calypso/blob/2380716c75c6bb05273b49e2d77a22f72f25fc56/client/my-sites/media-library/upload-button.scss) for this button here in `_components.scss`. This would move most of the CSS code to the `media` section, and fix this particular display issue.

jeremeylduvall commented:
@stephanethomas - Do you think you'll have a chance to review ^ in the upcoming days?

stephanethomas commented on assets/stylesheets/sections/post-editor.scss:
```diff
@@ -15,6 +15,7 @@
 @import '../main';
 @import '../shared/forms';
 
+@import './media';
```
The problem with importing the `media` section here is that it increases the size of the `post-editor` section proportionally, going from 145KB to 168KB. One way to address that problem would be to update our framework to actually support multiple [CSS bundles](](https://github.com/Automattic/wp-calypso/blob/master/client/wordpress-com.js#L233-L240)), e.g.:

```javascript
sections.push( {
  name: 'post-editor',
  paths: [ '/post', '/page', '/edit' ],
  module: 'post-editor',
  group: 'editor',
  secondary: true,
  css: ['post-editor', 'media']
} );
```

That's not something easy, but totally doable.

stephanethomas commented on assets/stylesheets/sections/media.scss:
```diff
@@ -0,0 +1,15 @@
+@import '../shared/mixins/heading';
+@import '../shared/mixins/long-content-fade';
+@import '../shared/mixins/noticon';
+
```
We're probably missing an import for the `z-index` function here since the `media.css` file generated contains rules like this:

```javascript
z-index: z-index("screen-reader-text-parent", ".screen-reader-text:focus");
```

jeremeylduvall commented:
@stephanethomas sorry this has taken me so long! I think we should be on the right track. Some recent updates:

1. I modified the bundler/utils.js so `getCssUrls` now can accept an array and returns an array regardless. For the post editor section then, we use `css: [ 'post-editor', 'media' ]`. `getCssUrls` now returns something like this:

```
ltr: [ url ],
rtl: [ url ]
```
2. The logic in `loadSectionCss` under `controller/shared.js` is updated as well to accept an array.
2. I added `@import '../shared/functions/z-index';` to `sections/media.scss`
3. I removed the bulk `@import 'my-sites/media-library/style';` into `_components.scss` and instead imported only `@import 'my-sites/media-library/upload-button';`

To test these recent changes, checkout this branch. Load http://calypso.localhost:3000/. I tested when they stylesheets were loaded in an incognito window. The media stylesheet is only loaded when you visit a /media route. Similarly, if you visit an editor route like `/page`, both the post editor and media stylesheets should load.

If desired, I can update all of the `css` declarations in `wordpress-com.js` to an array even for single strings. This would simplify some of the logic in `server/bundler/utils.js` since it would become very apparent that the function accepts an array instead of now where it accepts both a string or an array.

stephanethomas commented on client/controller/shared.js:
```diff
@@ -53,8 +53,18 @@ export function loadSectionCSS( context, next ) {
+		section.cssUrls.rtl.forEach( url => {
+			switchCSS( 'section-css', url, next );
+		} );
+
```
It looks like we could simplify this code a bit. What do you think of the following?

```javascript
const cssUrls = isRTL( context.store.getState() ) ? section.cssUrls.rtl : section.cssUrls.ltr;

cssUrls.forEach( cssUrl => {
  switchCSS( 'section-css', cssUrl, next );
} );

```

stephanethomas commented on server/bundler/utils.js:
```diff
@@ -43,9 +43,22 @@ function getHashedUrl( filename ) {
+	if ( Array.isArray( css ) ) {
+		css.map( cssSection => {
+			leftToRightCSS.push( getHashedUrl( 'sections/' + cssSection + '.css' ) );
+			rightToLeftCSS.push( getHashedUrl( 'sections-rtl/' + cssSection + '.css' ) );
```
We're missing the `.rtl` part in the file extension here.

stephanethomas commented on server/bundler/utils.js:
```diff
@@ -43,9 +43,22 @@ function getHashedUrl( filename ) {
+	} else {
+		leftToRightCSS.push( getHashedUrl( 'sections/' + css + '.css' ) );
+		rightToLeftCSS.push( getHashedUrl( 'sections-rtl/' + css + '.rtl.css' ) );
+	}
```
We could take this problem from the opposite angle in order to simplify that code, i.e. we could simply make sure we always deal with an array:

```javascript
const cssSections = Array.isArray( css ) ? css : [ css ];

cssSections.map( cssSection => {
  leftToRightCSS.push( getHashedUrl( 'sections/' + cssSection + '.css' ) );
  rightToLeftCSS.push( getHashedUrl( 'sections-rtl/' + cssSection + '.rtl.css' ) );
} );
```

That's something we often do in the API code by the way.

stephanethomas replied:
Thinking a bit more about it we could simplify that code further, and make it pure by introducing a new function:

```javascript
function getHashedUrls( cssSections, isRTL = false ) {
  return cssSections.map( ( cssSection ) => {
    const filename = isRTL ? `sections-rtl/${ cssSection }.rtl.css` : `sections/${ cssSection }.css`;

    return getHashedUrl( filename );
  } );
}

function getCssUrls( css ) {
  const cssSections = Array.isArray( css ) ? css : [ css ];

  return {
    ltr: getHashedUrls( cssSections ),
    rtl: getHashedUrls( cssSections, true )
  };
}
```

This is untested though :).

jeremeylduvall replied:
This makes total sense and it works well. Updated to use the second version and introduced new function.

jeremeylduvall commented:
@stephanethomas I fixed up `shared.js` and `utils.js` per your suggestions to make them simpler. I also rebased off master, and I'm not experiencing the display issues with RTL in testing. Can you take another look and let me know if you're still seeing those issues?

stephanethomas commented:
I tried opening the `Login` page but looking at the `Network` tab in the Chrome DevTools there was this strange request which failed:

```
http://calypso.localhost:3000/[%22/calypso/sections/login.css?v=7d0f843629%22]
```

Looking at the source of the page, it seems we have a problem here:

```html
<link rel="stylesheet" id="section-css" href="[&quot;/calypso/sections/login.css?v=7d0f843629&quot;]">
```

This only happens during server-side rendering; if you reload the page with Javascript disabled you'll see:

<img width="736" alt="screenshot" src="https://user-images.githubusercontent.com/594356/32776377-a2f762be-c932-11e7-9419-76d41a12d604.png">

stephanethomas commented:
@jeremeylduvall: I've added a commit to address the issue I've just mentioned. This fix is not perfect since it will only link to the first CSS bundle if a section uses more than one of them. It's not a problem for now though because we don't have any section rendered server-side with multiple CSS bundles yet. I'll correct that as part of https://github.com/Automattic/wp-calypso/pull/19396 anyway.

Note that I've rebased the pull request since I had to update my commit.

stephanethomas commented:
@jeremeylduvall: I've rebased again this pull request to remove a commit I added that fixed an issue with the `Media` menu item being not clickable with RTL languages. I suspect this fix requires a change to E2E tests, so I created https://github.com/Automattic/wp-calypso/pull/19790 to take care of that, and not block your pull request.

jeremeylduvall commented:
@stephanethomas thanks! I pulled down your commit and all looks okay to me. The e2e tests are failing at `Editor: Posts (mobile) Basic Public Post @canary @parallel @jetpack @chrome` specifically with `TimeoutError: Timed out waiting for element with css selector of '.editor-publish-button' to be clickable`. I wonder if that's related to this issue with failing editor mobile tests?

https://github.com/Automattic/wp-e2e-tests/issues/718

stephanethomas commented:
@jeremeylduvall: I've rebased again this branch against `master` to be compatible with Node 8. Let's see if we still have unit tests failing, or if it was an intermittent thing.

stephanethomas commented:
Unfortunately the unit test is still failing, this would require more investigation ...

jeremeylduvall commented:
Looking into this today/tomorrow!

Tug commented on client/controller/shared.js:
```diff
@@ -53,8 +53,12 @@ export function loadSectionCSS( context, next ) {
+		const cssUrls = isRTL( context.store.getState() ) ? section.cssUrls.rtl : section.cssUrls.ltr;
+
+		cssUrls.forEach( cssUrl => {
+			switchCSS( 'section-css', cssUrl, next );
```
We want to load the css here not switch, right?

stephanethomas commented:
@jeremeylduvall, I'd like to propose you to wait for https://github.com/Automattic/wp-calypso/pull/19396 to be merged before going any further with this one. Both pull requests are touching the exact same part of the code, which makes progressing in parallel hazardous.

jeremeylduvall commented:
Will do!

stephanethomas commented:
Thanks @jeremeylduvall for being so patient. https://github.com/Automattic/wp-calypso/pull/19396 was merged two weeks ago but I didn't have the time to circle back on your pull request. I tried to rebase it but during that process was quickly faced with the fact that those changes to support multiple CSS bundles will add a lot more complexity to our code now that we inject CSS bundles [with an id](https://github.com/Automattic/wp-calypso/pull/19396/commits/51ccc73208bf77d88460c7b1a4449a82c7597e85) into the page.

@Tug had some concerns about those particular changes when I discussed this pull request with him. His point was that this dependency should be expressed in the CSS bundle itself rather than in our [list of sections](https://github.com/Automattic/wp-calypso/blob/11c697d8663ed33de7b1b2ce1b0502a43411bf29/client/wordpress-com.js) (like you did originally). While I agree on the principle, I was worried that it affected the size of the `post-editor` section bundle since it basically means that we are including a CSS bundle into another one.

As mentioned [here](https://github.com/Automattic/wp-calypso/pull/18921#discussion_r145987929), we're increasing that CSS bundle from 147 KB to  171 KB approximately. This is unzipped (and unminified) but once zipped this only means an increase from 23 KB to 26 KB.

Given the minimal benefit, the fact that this might just be an edge case, the additional complexity incurred, and finally the perspective of Webpack 4 providing [first-class module support for CSS](https://medium.com/webpack/the-new-css-workflow-step-1-79583bd107d7) I think it makes indeed sense to ditch those changes that allowed multiple CSS bundles.

If that's fine to you, then I guess we only need to remove the last two commits from this pull request, add the `z-index` import (that was implemented in one of those commits), and rebase it. It will then be ready for a last round of reviews.

jeremeylduvall commented:
@stephanethomas I believe this should be ready for review. Just to recap, the main things now included in the commit are:

- Pulled out separate css bundle for the `/media` route (including z-index)
- Added the styles for `my-sites/media-library/upload-button` to `_components.scss` to avoid breaking "Add" button
- Added `./media` import to `post-editor` bundle

I tested both .rtl and .ltr languages to make sure everything was loading correctly. I did find one small issue with .rtl that I believe is related to [the race condition mentioned here](https://github.com/Automattic/wp-calypso/pull/18921#pullrequestreview-75467885).

<img width="431" alt="screen shot 2017-12-06 at 7 37 34 am" src="https://user-images.githubusercontent.com/7240478/33666813-5732d156-da58-11e7-924a-da3647af64e5.png">

It only happens every once in awhile. Per your comment, I don't believe that will be a factor in production, correct?

Tug commented:
LGTM 👍 

I was able to reproduce the race condition but well it happens on master/production as well. It seems to be a caching issue as it will disappear after a some time has passed and a reload.
